### ISSUE TITLE
Memory over-read in L1TMuonOverlapTrackProducer

### ISSUE THREAD

**Issue by Dr15Jones** [cms-sw/cmssw#25785]
The address sanitizer caught a memory out of bounds read in `GhostBusterPreferRefDt::select` which is being called by `L1TMuonOverlapTrackProducer`.

**Comment by Dr15Jones** [comment]
The link to the log is:
https://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc7_amd64_gcc700/CMSSW_10_5_ASAN_X_2019-01-25-2300/pyRelValMatrixLogs/run/140.56_RunHI2018+RunHI2018+RECOHID18+HARVESTDHI18/step2_RunHI2018+RunHI2018+RECOHID18+HARVESTDHI18.log#/

with the relevant information being:
```
    #0 0x7f8a83d17cb7 in void std::__insertion_sort<std::reverse_iterator<__gnu_cxx::__normal_iterator<AlgoMuon*, std::vector<AlgoMuon, std::allocator<AlgoMuon> > > >, __gnu_cxx::__ops::_Iter_comp_iter<GhostBusterPreferRefDt::select(std::vector<AlgoMuon, std::allocator<AlgoMuon> >, int)::{lambda(AlgoMuon const&, AlgoMuon const&)#1}> >(std::reverse_iterator<__gnu_cxx::__normal_iterator<AlgoMuon*, std::vector<AlgoMuon, std::allocator<AlgoMuon> > > >, __gnu_cxx::__ops::_Iter_comp_iter<GhostBusterPreferRefDt::select(std::vector<AlgoMuon, std::allocator<AlgoMuon> >, int)::{lambda(AlgoMuon const&, AlgoMuon const&)#1}>, __gnu_cxx::__ops::_Iter_comp_iter<GhostBusterPreferRefDt::select(std::vector<AlgoMuon, std::allocator<AlgoMuon> >, int)::{lambda(AlgoMuon const&, AlgoMuon const&)#1}>) (/cvmfs/cms-ib.cern.ch/nweek-02560/slc7_amd64_gcc700/cms/cmssw/CMSSW_10_5_ASAN_X_2019-01-25-2300/lib/slc7_amd64_gcc700/libL1TriggerL1TMuonOverlap.so+0x23cb7)
    #1 0x7f8a83d253f3 in GhostBusterPreferRefDt::select(std::vector<AlgoMuon, std::allocator<AlgoMuon> >, int) (/cvmfs/cms-ib.cern.ch/nweek-02560/slc7_amd64_gcc700/cms/cmssw/CMSSW_10_5_ASAN_X_2019-01-25-2300/lib/slc7_amd64_gcc700/libL1TriggerL1TMuonOverlap.so+0x313f3)
    #2 0x7f8a83d6805d in OMTFReconstruction::getProcessorCandidates(unsigned int, l1t::tftype, int, BXVector<l1t::RegionalMuonCand>&) (/cvmfs/cms-ib.cern.ch/nweek-02560/slc7_amd64_gcc700/cms/cmssw/CMSSW_10_5_ASAN_X_2019-01-25-2300/lib/slc7_amd64_gcc700/libL1TriggerL1TMuonOverlap.so+0x7405d)
    #3 0x7f8a83d69c8d in OMTFReconstruction::reconstruct(edm::Event const&, edm::EventSetup const&) (/cvmfs/cms-ib.cern.ch/nweek-02560/slc7_amd64_gcc700/cms/cmssw/CMSSW_10_5_ASAN_X_2019-01-25-2300/lib/slc7_amd64_gcc700/libL1TriggerL1TMuonOverlap.so+0x75c8d)
    #4 0x7f8a83ea1648 in L1TMuonOverlapTrackProducer::produce(edm::Event&, edm::EventSetup const&) (/cvmfs/cms-ib.cern.ch/nweek-02560/slc7_amd64_gcc700/cms/cmssw/CMSSW_10_5_ASAN_X_2019-01-25-2300/lib/slc7_amd64_gcc700/pluginL1TriggerL1TMuonOverlapPlugins.so+0x2b648)
    #5 0x7f8affa1cd94 in edm::one::EDProducerBase::doEvent(edm::EventPrincipal const&, edm::EventSetup const&, edm::ActivityRegistry*, edm::ModuleCallingContext const*) (/cvmfs/cms-ib.cern.ch/nweek-02560/slc7_amd64_gcc700/cms/cmssw/CMSSW_10_5_ASAN_X_2019-01-25-2300/lib/slc7_amd64_gcc700/libFWCoreFramework.so+0x74dd94)
    #6 0x7f8aff9c05fc in edm::WorkerT<edm::one::EDProducerBase>::implDo(edm::EventPrincipal const&, edm::EventSetup const&, edm::ModuleCallingContext const*) (/cvmfs/cms-ib.cern.ch/nweek-02560/slc7_amd64_gcc700/cms/cmssw/CMSSW_10_5_ASAN_X_2019-01-25-2300/lib/slc7_amd64_gcc700/libFWCoreFramework.so+0x6f15fc)
    #7 0x7f8aff44b2c3 in decltype ({parm#1}()) edm::convertException::wrap<bool edm::Worker::runModule<edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1> >(edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1>::MyPrincipal const&, edm::EventSetup const&, edm::StreamID, edm::ParentContext const&, edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1>::Context const*)::{lambda()#1}>(bool edm::Worker::runModule<edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1> >(edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1>::MyPrincipal const&, edm::EventSetup const&, edm::StreamID, edm::ParentContext const&, edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1>::Context const*)::{lambda()#1}) (/cvmfs/cms-ib.cern.ch/nweek-02560/slc7_amd64_gcc700/cms/cmssw/CMSSW_10_5_ASAN_X_2019-01-25-2300/lib/slc7_amd64_gcc700/libFWCoreFramework.so+0x17c2c3)
    #8 0x7f8aff44b752 in bool edm::Worker::runModule<edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1> >(edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1>::MyPrincipal const&, edm::EventSetup const&, edm::StreamID, edm::ParentContext const&, edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1>::Context const*) (/cvmfs/cms-ib.cern.ch/nweek-02560/slc7_amd64_gcc700/cms/cmssw/CMSSW_10_5_ASAN_X_2019-01-25-2300/lib/slc7_amd64_gcc700/libFWCoreFramework.so+0x17c752)
    #9 0x7f8aff44c21c in std::__exception_ptr::exception_ptr edm::Worker::runModuleAfterAsyncPrefetch<edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1> >(std::__exception_ptr::exception_ptr const*, edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1>::MyPrincipal const&, edm::EventSetup const&, edm::StreamID, edm::ParentContext const&, edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1>::Context const*) (/cvmfs/cms-ib.cern.ch/nweek-02560/slc7_amd64_gcc700/cms/cmssw/CMSSW_10_5_ASAN_X_2019-01-25-2300/lib/slc7_amd64_gcc700/libFWCoreFramework.so+0x17d21c)
    #10 0x7f8aff457216 in void edm::SerialTaskQueueChain::actionToRun<edm::Worker::RunModuleTask<edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1> >::execute()::{lambda()#1}&>(edm::Worker::RunModuleTask<edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1> >::execute()::{lambda()#1}&) (/cvmfs/cms-ib.cern.ch/nweek-02560/slc7_amd64_gcc700/cms/cmssw/CMSSW_10_5_ASAN_X_2019-01-25-2300/lib/slc7_amd64_gcc700/libFWCoreFramework.so+0x188216)
    #11 0x7f8aff457d9c in edm::SerialTaskQueue::QueuedTask<void edm::SerialTaskQueueChain::push<edm::Worker::RunModuleTask<edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1> >::execute()::{lambda()#1}&>(edm::Worker::RunModuleTask<edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1> >::execute()::{lambda()#1}&)::{lambda()#1}>::execute() (/cvmfs/cms-ib.cern.ch/nweek-02560/slc7_amd64_gcc700/cms/cmssw/CMSSW_10_5_ASAN_X_2019-01-25-2300/lib/slc7_amd64_gcc700/libFWCoreFramework.so+0x188d9c)
    #12 0x7f8afd47d175 in tbb::internal::custom_scheduler<tbb::internal::IntelSchedulerTraits>::local_wait_for_all(tbb::task&, tbb::task*) ../../src/tbb/custom_scheduler.h:521
    #13 0x7f8afd476a5f in tbb::internal::arena::process(tbb::internal::generic_scheduler&) ../../src/tbb/arena.cpp:160
    #14 0x7f8afd4753e2 in tbb::internal::market::process(rml::job&) ../../src/tbb/market.cpp:710
    #15 0x7f8afd4715e2 in tbb::internal::rml::private_worker::run() ../../src/tbb/private_server.cpp:270
    #16 0x7f8afd471848 in tbb::internal::rml::private_worker::thread_routine(void*) ../../src/tbb/private_server.cpp:223
    #17 0x7f8afc13ce24 in start_thread (/lib64/libpthread.so.0+0x7e24)
    #18 0x7f8afbe66bac in __clone (/lib64/libc.so.6+0xfebac)

0x603004e8d87c is located 4 bytes to the left of 32-byte region [0x603004e8d880,0x603004e8d8a0)
allocated by thread T8 here:
    #0 0x7f8affc6e1e0 in operator new(unsigned long) ../../../../libsanitizer/asan/asan_new_delete.cc:80
    #1 0x7f8afe1e4811 in std::vector<int, std::allocator<int> >::_M_default_append(unsigned long) (/cvmfs/cms-ib.cern.ch/nweek-02560/slc7_amd64_gcc700/cms/cmssw/CMSSW_10_5_ASAN_X_2019-01-25-2300/external/slc7_amd64_gcc700/lib/libRIO.so+0x1bd811)

Thread T7 created by T0 here:
    #0 0x7f8affbca230 in __interceptor_pthread_create ../../../../libsanitizer/asan/asan_interceptors.cc:243
    #1 0x7f8afd4714a1 in rml::internal::thread_monitor::launch(void* (*)(void*), void*, unsigned long) ../../src/tbb/../rml/server/thread_monitor.h:221
    #2 0x7f8afd4714a1 in tbb::internal::rml::private_worker::wake_or_launch() ../../src/tbb/private_server.cpp:301
    #3 0x7f8afd4714a1 in tbb::internal::rml::private_server::wake_some(int) ../../src/tbb/private_server.cpp:399
    #4 0x60c00080643f  (<unknown module>)

Thread T8 created by T0 here:
    #0 0x7f8affbca230 in __interceptor_pthread_create ../../../../libsanitizer/asan/asan_interceptors.cc:243
    #1 0x7f8afd4714a1 in rml::internal::thread_monitor::launch(void* (*)(void*), void*, unsigned long) ../../src/tbb/../rml/server/thread_monitor.h:221
    #2 0x7f8afd4714a1 in tbb::internal::rml::private_worker::wake_or_launch() ../../src/tbb/private_server.cpp:301
    #3 0x7f8afd4714a1 in tbb::internal::rml::private_server::wake_some(int) ../../src/tbb/private_server.cpp:399
    #4 0x7f8af74080ff  (<unknown module>)

SUMMARY: AddressSanitizer: heap-buffer-overflow (/cvmfs/cms-ib.cern.ch/nweek-02560/slc7_amd64_gcc700/cms/cmssw/CMSSW_10_5_ASAN_X_2019-01-25-2300/lib/slc7_amd64_gcc700/libL1TriggerL1TMuonOverlap.so+0x23cb7) in void std::__insertion_sort<std::reverse_iterator<__gnu_cxx::__normal_iterator<AlgoMuon*, std::vector<AlgoMuon, std::allocator<AlgoMuon> > > >, __gnu_cxx::__ops::_Iter_comp_iter<GhostBusterPreferRefDt::select(std::vector<AlgoMuon, std::allocator<AlgoMuon> >, int)::{lambda(AlgoMuon const&, AlgoMuon const&)#1}> >(std::reverse_iterator<__gnu_cxx::__normal_iterator<AlgoMuon*, std::vector<AlgoMuon, std::allocator<AlgoMuon> > > >, __gnu_cxx::__ops::_Iter_comp_iter<GhostBusterPreferRefDt::select(std::vector<AlgoMuon, std::allocator<AlgoMuon> >, int)::{lambda(AlgoMuon const&, AlgoMuon const&)#1}>, __gnu_cxx::__ops::_Iter_comp_iter<GhostBusterPreferRefDt::select(std::vector<AlgoMuon, std::allocator<AlgoMuon> >, int)::{lambda(AlgoMuon const&, AlgoMuon const&)#1}>)
Shadow bytes around the buggy address:
  0x0c06809c9ab0: 00 00 00 00 fa fa 00 00 00 00 fa fa 00 00 00 00
  0x0c06809c9ac0: fa fa 00 00 00 00 fa fa 00 00 00 00 fa fa 00 00
  0x0c06809c9ad0: 00 00 fa fa 00 00 00 00 fa fa 00 00 00 fa fa fa
  0x0c06809c9ae0: 00 00 00 00 fa fa 00 00 00 00 fa fa 00 00 00 00
  0x0c06809c9af0: fa fa 00 00 00 07 fa fa 00 00 00 00 fa fa 00 00
=>0x0c06809c9b00: 00 07 fa fa 00 00 00 00 fa fa 00 00 00 00 fa[fa]
  0x0c06809c9b10: 00 00 00 00 fa fa 00 00 00 00 fa fa fa fa fa fa
  0x0c06809c9b20: fa fa 00 00 03 fa fa fa 00 00 00 00 fa fa 00 00
  0x0c06809c9b30: 00 00 fa fa 00 00 00 00 fa fa 00 00 00 fa fa fa
  0x0c06809c9b40: 00 00 00 fa fa fa 00 00 00 fa fa fa 00 00 00 00
  0x0c06809c9b50: fa fa 00 00 00 07 fa fa 00 00 00 fa fa fa 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==10563==ABORTING
```

**Comment by Dr15Jones** [comment]
Looking at the code, the problem is likely 
https://github.com/cms-sw/cmssw/blob/4626d53e6562a9fe9f18f25a9f48d7cc858cb0fd/L1Trigger/L1TMuonOverlap/src/GhostBusterPreferRefDt.cc#L23
or
https://github.com/cms-sw/cmssw/blob/4626d53e6562a9fe9f18f25a9f48d7cc858cb0fd/L1Trigger/L1TMuonOverlap/src/GhostBusterPreferRefDt.cc#L24

with either `a.getRefLayer()` or `b.getRefLayer()` returning a value outside of the bounds of the vector. It is likely that one of these values is less than 0 given the access violation is "to the left of" the container.

**Comment by Dr15Jones** [comment]
And `getRefLayer()` can indeed return -1 since that is the value for the defaultly constructed `AlgoMuon` and is the value `OMTFSorter::sortSingleResult` returns if there were no hits:

https://github.com/cms-sw/cmssw/blob/4626d53e6562a9fe9f18f25a9f48d7cc858cb0fd/L1Trigger/L1TMuonOverlap/src/OMTFSorter.cc#L32-L42

**Comment by Dr15Jones** [comment]
assign l1

**Comment by Dr15Jones** [comment]
What is the timeline for getting this fixed?

**Comment by rekovic** [comment]
@Dr15Jones 
Next week.

**Comment by Dr15Jones** [comment]
How goes the progress on this issue?

**Comment by Dr15Jones** [comment]
Is there progress being made in fixing this issue?

**Comment by Dr15Jones** [comment]
@fabiocos not having this fixed is preventing us from finding potentially other problems using ASAN (since jobs stop at the first problem found).

**Comment by rekovic** [comment]
@Dr15Jones @fabiocos 
I am so sorry.   I was not getting the automated e-mail into my inbox (now fixed), and I completely forgot about this issue.  Looking into this this afternoon.

**Comment by fabiocos** [comment]
this is the output of a valgrind cross check, that goes in the direction of the diagnosis of @Dr15Jones:

```
<error>
  <unique>0xb88ba</unique>
  <tid>1</tid>
  <kind>InvalidRead</kind>
  <what>Invalid read of size 4</what>
  <stack>
    <frame>
      <ip>0x228A1A787</ip>
      <obj>/build/fabiocos/106X/l1t/CMSSW_10_6_X_2019-03-17-0000/lib/slc7_amd64_gcc700/libL1TriggerL1TMuonOverlap.so</obj>
      <fn>void std::__insertion_sort&lt;std::reverse_iterator&lt;__gnu_cxx::__normal_iterator&lt;AlgoMuon*, std::vector&lt;AlgoMuon, std::allocator&lt;Alg
oMuon&gt; &gt; &gt; &gt;, __gnu_cxx::__ops::_Iter_comp_iter&lt;GhostBusterPreferRefDt::select(std::vector&lt;AlgoMuon, std::allocator&lt;AlgoMuon&gt; &gt;
, int)::{lambda(AlgoMuon const&amp;, AlgoMuon const&amp;)#1}&gt; &gt;(std::reverse_iterator&lt;__gnu_cxx::__normal_iterator&lt;AlgoMuon*, std::vector&lt;A
lgoMuon, std::allocator&lt;AlgoMuon&gt; &gt; &gt; &gt;, __gnu_cxx::__ops::_Iter_comp_iter&lt;GhostBusterPreferRefDt::select(std::vector&lt;AlgoMuon, std::
allocator&lt;AlgoMuon&gt; &gt;, int)::{lambda(AlgoMuon const&amp;, AlgoMuon const&amp;)#1}&gt;, __gnu_cxx::__ops::_Iter_comp_iter&lt;GhostBusterPreferRefD
t::select(std::vector&lt;AlgoMuon, std::allocator&lt;AlgoMuon&gt; &gt;, int)::{lambda(AlgoMuon const&amp;, AlgoMuon const&amp;)#1}&gt;)</fn>
    </frame>
    <frame>
      <ip>0x228A1B8D9</ip>
      <obj>/build/fabiocos/106X/l1t/CMSSW_10_6_X_2019-03-17-0000/lib/slc7_amd64_gcc700/libL1TriggerL1TMuonOverlap.so</obj>
      <fn>GhostBusterPreferRefDt::select(std::vector&lt;AlgoMuon, std::allocator&lt;AlgoMuon&gt; &gt;, int)</fn>
    </frame>
    <frame>
      <ip>0x228A2EFD3</ip>
      <obj>/build/fabiocos/106X/l1t/CMSSW_10_6_X_2019-03-17-0000/lib/slc7_amd64_gcc700/libL1TriggerL1TMuonOverlap.so</obj>
      <fn>OMTFReconstruction::getProcessorCandidates(unsigned int, l1t::tftype, int, BXVector&lt;l1t::RegionalMuonCand&gt;&amp;)</fn>
    </frame>
    <frame>
      <ip>0x228A2FC1D</ip>
      <obj>/build/fabiocos/106X/l1t/CMSSW_10_6_X_2019-03-17-0000/lib/slc7_amd64_gcc700/libL1TriggerL1TMuonOverlap.so</obj>
      <fn>OMTFReconstruction::reconstruct(edm::Event const&amp;, edm::EventSetup const&amp;)</fn>
    </frame>
    <frame>
      <ip>0x2289D8C98</ip>
      <obj>/build/fabiocos/106X/l1t/CMSSW_10_6_X_2019-03-17-0000/lib/slc7_amd64_gcc700/pluginL1TriggerL1TMuonOverlapPlugins.so</obj>
      <fn>L1TMuonOverlapTrackProducer::produce(edm::Event&amp;, edm::EventSetup const&amp;)</fn>
    </frame>
    <frame>
      <ip>0x4C9558A</ip>
      <obj>/cvmfs/cms-ib.cern.ch/nweek-02568/slc7_amd64_gcc700/cms/cmssw/CMSSW_10_6_X_2019-03-17-0000/lib/slc7_amd64_gcc700/libFWCoreFramework.so</obj>
      <fn>edm::one::EDProducerBase::doEvent(edm::EventPrincipal const&amp;, edm::EventSetupImpl const&amp;, edm::ActivityRegistry*, edm::ModuleCallingCont
ext const*)</fn>
    </frame>
    <frame>
```

From the software sanity point of view the minimal solution seems to put a protection against negative refLayers to be used as vector index. But L1T experts should evaluate why this situation is present, and what is the best action the code should take to manage it.

**Comment by rekovic** [comment]
@Dr15Jones 
I tried to look into the log file you point out above, but I cannot access it anymore, as it is long past.  Which exact valgrind command did you use to reproduce this.  
I provided a temporary fix, which should solve the problem but would be good to use your command and check for any other issues.  Thanks.

**Comment by Dr15Jones** [comment]
@rekovic I never ran valgrind, I just looked in the nightly builds output for ASAN. See

https://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc7_amd64_gcc700/CMSSW_10_6_ASAN_X_2019-03-15-2300/pyRelValMatrixLogs/run/140.56_RunHI2018+RunHI2018+RECOHID18+HARVESTDHI18/step2_RunHI2018+RunHI2018+RECOHID18+HARVESTDHI18.log#/

which was created this weekend.

So to replay, create a CMSSW_10_6_ASAN_X_2019-03-15-2300 work area and then do runTheMatrix.py -l 140.56 -t 4

**Comment by fabiocos** [comment]
@Dr15Jones @rekovic I have run valgrind on the step2 of 140.56, after creating it with the regular runTheMatrix.py -l 140.56 

valgrind --tool=memcheck --suppressions=$CMSSW_RELEASE_BASE/src/Utilities/ReleaseScripts/data/cms-valgrind-memcheck.supp --num-callers=20 --xml=yes --xml-file=valgrind-%p.xml --smc-check=all-non-file --track-origins=yes --leak-check=yes --freelist-vol=10000000000 --undef-value-errors=yes --expensive-definedness-checks=yes --malloc-fill=0xa5 --free-fill=0x5a cmsRunGlibC step2_RAW2DIGI_L1Reco_RECO_ALCA_SKIM_EI_DQM.py

**Comment by fabiocos** [comment]
@Dr15Jones @rekovic rerunning valgrind after #26208 has been merged, the problem looks disappeared

**Comment by fabiocos** [comment]
@Dr15Jones in the latest ASAN build workflow 140.56 looks successful, I would consider this issue as solved. @rekovic still experts should ensure that the added protection is the correct/best solution from the physics point of view (unless already done)

**Comment by Dr15Jones** [comment]
@fabiocos the issue is assigned to l1, not core. I thought the normal procedure is the the assigned party to sign and then someone with repository access closes the issue.

**Comment by rekovic** [comment]
@fabiocos 
From the physics performance point of view the provided fix in #26208 has been confirmed as satisfactory by the OMTF experts. 

The issue can be closed.

**Comment by rekovic** [comment]
+1
